doc/user-manual/language/polarity.lagda.rst

```diff
@@ -0,0 +1,109 @@
+annotations: ::
+
+  strictly-positive : @++ Set → Set
+  strictly-positive A = Nat → A
+
+  positive : @+ Set → Set
+  positive A = (A → Nat) → A
+
+  negative : @- Set → Set
+  negative A = A → Nat
+
+  mixed : @mixed Set → Set
+  mixed A = A → A
+
+  unused : @unused Set → Set
+  unused A = Nat
+
+The standard example use-case is defining the fix-point of any strictly positive
+type-former: ::
+
+  data Mu (F : @++ Set → Set) : Set where
+    fix : F (Mu F) → Mu F
+
+In the example above, because ``F`` is specified as using its argument strictly
+positively, the positivity checker allows ``F (Mu F)`` as an argument to a
+constructor since it knows the recursive call to ``Mu F`` is in strictly
+positive position.
+
+When defining functions that take arguments annotated with ``@++``, the typing
+rules ensure that those arguments may never actually appear to the left of an
+arrow. They can syntactically appear to the left of an arrow as arguments to
+functions that don't use their arguments, such as in the correct example below:
+::
+
+  const : @unused Set → Set
+  const _ = Nat
+
+  typechecks : @++ Set → Set
+  typechecks A = const A → Nat
+
+The polarity modality
+=====================
+
+Agda implements polarity annotations using a modal system. Here are the
+different modalities and their meaning:
+
++--------+-------------------+-------------------------------------+
+|Notation|Name               |Possible use                         |
++========+===================+=====================================+
+|@++     |Strictly positive  |Anywhere except in the domain of a   |
+|        |                   |pi/function type                     |
++--------+-------------------+-------------------------------------+
+|@+      |Positive           |In an even number of nested domains  |
+|        |                   |of pi/function types                 |
++--------+-------------------+-------------------------------------+
+|@-      |Negative           |In an odd number of nested domains of|
+|        |                   |pi/function types                    |
++--------+-------------------+-------------------------------------+
+|@mixed  |Mixed              |Anywhere                             |
++--------+-------------------+-------------------------------------+
+|@unused |Unused             |Nowhere                              |
++--------+-------------------+-------------------------------------+
+
+When Agda type-checks any definition, it ensures that the variables bound by a
+lambda-abstraction with annotated types satisfy those restrictions.  Note that
+there is no such restriction when checking the body of a pi type, so for example
+``(@++ A : Set) → (A → A)`` is perfectly valid!
+
+Polarity annotations can only appear on domains of function types and
+data/record type parameters. Pattern-matching on annotated arguments is only
+supported for mixed arguments (but polarities are only useful when annotating
+``Set`` anyways).
```

Comment:
Could you expand a bit on this?

Reply:
The conversion checker seems to ignore arguments that are known to be unused:
```agda
open import Agda.Builtin.Bool
open import Agda.Builtin.Equality

postulate
  f : Bool → Bool

{-# POLARITY f _ #-}

_ : f true ≡ f false
_ = refl
```
Please verify that this is the case also when polarity annotations are used, add a test case, and modify the text above.

Reply:
Is this even desirable? We have irrelevance for that.

Reply:
Yeah, I also think it's better to have "unused" just mean "neither positive nor negative" but not have it simulate any form of irrelevance.

Reply:
Neither positive nor negative is the mixed modality. The meaning of unused is somewhere in the interval ranging from parametric/natural to irrelevant.

Reply:
I thought mixed means that you can use both positive and negative arguments, so it is "both positive and negative"?

Reply:
Ah, this is a matter of perspective. Yes, a mixed variable can be used both in positive and negative positions. The result is a function that is neither positive (a covariant functor) nor negative (a contravariant functor).

Reply:
I'm not a big fan of this feature, but I don't think that polarity pragmas and polarity annotations should have different semantics.

The Agda 2.3.2 changelog contains some motivation for the feature (which was [implemented](https://github.com/agda/agda/commit/41e59c645710fb5a8dad8d3ad2395a0539fc6910) by @andreasabel):

https://github.com/agda/agda/blob/b906afbf13a98995bc6901e0cf48a3f96059ca70/doc/release-notes/2.3.2.md?plain=1#L480-L572

Reply:
I was not aware that the unused polarity pragma interacted in this way with the conversion machinery. Ideally, our `@unused` should be syntactically equivalent to `@irrelevant`, but we chose to keep them separate to simplify the base implementation, and to avoid having to deal with all the problems this additional feature could bring.

Reply:
Now that I'm looking over the review, do we actually want to add this? I think I can implement it without too many problems, but I wanted to keep the implementation as simple as possible, only adding the necessary polarity stuff. Is it strictly needed?

Reply:
I'm not sure what "this" refers to.

Reply:
See #6439.